tsl/src/planner.c

```diff
@@ -179,6 +180,34 @@ copy_mode_enabled(void)
 	return strcmp(enable_copy, "true") == 0;
 }
 
+/* Query tree walker to examine RTE contents for a distributed hypertable */
+static bool
+distributed_rtes_walker(Node *node, bool *isdistributed)
+{
+	/* if isdistributed is already set, then no need to walk further */
+	if (*isdistributed)
+		return true;
+	if (node == NULL)
+		return false;
+	if (IsA(node, RangeTblEntry))
+	{
+		RangeTblEntry *rte = (RangeTblEntry *) node;
+
+		if (rte->rtekind == RTE_RELATION)
+			ts_rte_is_hypertable(rte, isdistributed);
```

Comment:
We could potentially do better here by checking whether the other distributed hypertable includes the same data nodes. If the set of data nodes is disjoint, the INSERT ... SELECT could still be allowed. 

However, this is not a very common case, but it might be worth at least leaving a comment about it for the future.

Reply:
Added a comment.